### PR TITLE
Track source for objects and fields with [synthetic_source_keep:arrays] in arrays as ignored

### LINKED ISSUE #115261: Incorrect synthetic source for object arrays withtin nested objects
For object arrays that are within nested object and contain leaf values that are "mixed" (some synthetic source values come from doc_values and some from ignored source) synthetic source is not correct.

Following test demonstrates that:
```
    public void testObjectArrayWithinNestedObjects() throws IOException {
        DocumentMapper documentMapper = createMapperService(syntheticSourceMapping(b -> {
            b.startObject("path").startObject("properties");
            {
                b.startObject("to").field("type", "nested").startObject("properties");
                {
                    b.startObject("obj").startObject("properties");
                    {
                        b.startObject("id").field("type", "integer").field("synthetic_source_keep", "arrays").endObject();
                    }
                    b.endObject().endObject();
                }
                b.endObject().endObject();
            }
            b.endObject().endObject();
        })).documentMapper();

        var syntheticSource = syntheticSource(documentMapper, b -> {
            b.startObject("path");
            {
                b.startObject("to");
                {
                    b.startArray("obj");
                    {
                        b.startObject().array("id", 1, 20, 3).endObject();
                        b.startObject().field("id", 10).endObject();
                    }
                    b.endArray();
                }
                b.endObject();
            }
            b.endObject();
        });
        assertEquals("""
            {"path":{"to":{"obj":{"id":[1,20,3,10]}}}}""", syntheticSource);
    }
```

This test produces `{"path":{"to":{"obj":{"id":[1,20,3]}}}}`.

### PR DESCRIPTION
**Pull request by kkrik-es**
Objects and fields with `[synthetic_source_keep:arrays]`, when nested under object arrays with [synthetic_source_keep:none], are the only use cases requiring the second parsing pass for docs with synthetic source. It's simpler to just record singleton values for them in ignored source, avoiding the extra complexity and performance hit for the second pass.

One downside is that there's additional overhead when the array contains only singleton instances of these objects and fields - using ignored source is not needed here. Still, this should be a corner case (parent object not tracking source, sub-fields tracking array source, appearing as singletons only), so this looks like a good tradeoff.

Fixes #115261

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 1]
```diff
@@ -653,13 +653,6 @@ public Iterator<Setting<?>> settings() {
         Property.Final
     );
 
-    public static final Setting<Boolean> SYNTHETIC_SOURCE_SECOND_DOC_PARSING_PASS_SETTING = Setting.boolSetting(
-        "index.synthetic_source.enable_second_doc_parsing_pass",
-        true,
-        Property.IndexScope,
-        Property.Dynamic
-    );
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:661]: Not sure if it's safe to remove a setting, would this cause issues with rolling upgrades? It shouldn't be set anywhere, of course.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:661]: Shard of indices using setting would not be able allocate if this setting is used, but it has been removed. That what branches did this setting get back ported? I think just removing the setting, should fix the issue, but would be good to double check this.

**Review by martijnvg — COMMENTED** [review]
This looks like a good simplification that should also improve performance. Did you have a chance to run this with the elastic/logs track and see what how this improves indexing throughput?

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java (hunk 1 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:661]: This was added to 8.16, not used anywhere as it's a last-resort setting: https://github.com/elastic/elasticsearch/pull/114649

**Comment by kkrik-es** [comment]
> This looks like a good simplification that should also improve performance. Did you have a chance to run this with the elastic/logs track and see what how this improves indexing throughput?

I don't think this will show up in benchmarks, as we've enabled storing array source in logsdb.. The second pass only kicks in if the parent array is not stored.

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:661]: I see. I'm not against removing this. I was trying to quantify the severity of the issue if someone ends up using this setting and then upgrades.

**Review by lkts — APPROVED** [review]

**Comment by lkts** [comment]
Nice!

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**